File: ml-utils/src/main/java/datasets/DataSet.java
package datasets;

import java.util.*;

public class DataSet<T extends Instance> implements Iterable<T> {
    private T[] instances;
    private boolean hasDiscreteOutput;

    public DataSet(T[] instances, boolean hasDiscreteOutput) {
        this.instances = instances;
        this.hasDiscreteOutput = hasDiscreteOutput;
    }

    public Instance[] getInstances() {
        return instances;
    }

    public List[] splitDataSetRandomly(int ways) {
        List[] divisions = new List[ways];

        for (int i = 0; i < divisions.length; i++) {
            divisions[i] = new ArrayList<T>();
        }

        List<T> instanceList = new ArrayList<>(Arrays.asList(instances));

        for (int i = 0; i < instances.length; i++) {
            divisions[i % divisions.length].add(instanceList.remove((int)(Math.random() * instanceList.size())));
        }

        return divisions;
    }

    @Override
    public Iterator<T> iterator() {
        return new DataSetIterator();
    }

    public boolean hasDiscreteOutput() {
        return hasDiscreteOutput;
    }

    private class DataSetIterator implements Iterator<T> {

        private int index;

        @Override
        public boolean hasNext() {
            return index < instances.length;
        }

        @Override
        public T next() {
            return (T) instances[index++];
        }
    }
}

File: ml-utils/src/main/java/datasets/DataSet.kt
package datasets

import util.Csv
import java.util.*

class DataSet<T : Instance>(private val instances: Array<T>, private val hasDiscreteOutput: Boolean) : Iterable<T> {

    fun getInstances(): Array<T> {
        return instances
    }

    fun splitDataSetRandomly(ways: Int): Array<List<T>> {
        val divisions = Array(ways, { mutableListOf<T>()})

        val instanceList = ArrayList(Arrays.asList(*instances))

        for (i in instances.indices) {
            divisions[i % divisions.size].add(instanceList.removeAt((Math.random() * instanceList.size).toInt()))
        }

        return divisions.toList().toTypedArray()
    }

    override fun iterator(): Iterator<T> {
        return DataSetIterator()
    }

    fun hasDiscreteOutput(): Boolean {
        return hasDiscreteOutput
    }

    override fun toString(): String {
        val numInputs = instances[0].input.size

        val inputColumnNames = (1..numInputs).map { i -> "Input${i}" }.toTypedArray()

        val csv = Csv(*inputColumnNames, "Output")

        instances.forEach { csv.addRow(*it.input.toTypedArray(), it.output) }

        return csv.toString()
    }

    fun deepCopy(): DataSet<T> {

        instances.map {  }

        return DataSet(instances, hasDiscreteOutput)
    }

    private inner class DataSetIterator : Iterator<T> {

        private var index: Int = 0

        override fun hasNext(): Boolean {
            return index < instances.size
        }

        override fun next(): T {
            return instances[index++]
        }
    }
}
